parachain/pallets/control/src/lib.rs

```diff
@@ -131,33 +155,147 @@ pub mod pallet {
+			let origin_location: MultiLocation = T::ControlOrigin::ensure_origin(origin)?;
 
-			// Normalize all locations relative to the relay chain unless its the relay itself.
-			let relay_location = T::RelayLocation::get();
-			if location != relay_location {
-				location
-					.reanchor(&relay_location, T::UniversalLocation::get())
-					.or(Err(Error::<T>::LocationConversionFailed))?;
-			}
+			let (agent_id, _, location) = Self::convert_location(origin_location)?;
 
-			// Hash the location to produce an agent id
-			let agent_id = T::AgentHashedDescription::convert_location(&location)
-				.ok_or(Error::<T>::LocationConversionFailed)?;
+			log::debug!(
+				target: LOG_TARGET,
+				"💫 Create Agent request with agent_id {:?}, origin_location at {:?}, location at {:?}",
+				agent_id,
+				origin_location,
+				location
+			);
 
 			// Record the agent id or fail if it has already been created
-			if Agents::<T>::get(agent_id).is_some() {
-				return Err(Error::<T>::AgentAlreadyCreated.into())
-			}
+			ensure!(!Agents::<T>::contains_key(agent_id), Error::<T>::AgentAlreadyCreated);
+
 			Agents::<T>::insert(agent_id, ());
 
 			let message =
 				Message { origin: T::OwnParaId::get(), command: Command::CreateAgent { agent_id } };
-			Self::submit_outbound(message)?;
+			Self::submit_outbound(message.clone())?;
+
+			log::debug!(
+				target: LOG_TARGET,
+				"💫 Create Agent request processed with outbound message {:?}",
+				message
+			);
 
 			Self::deposit_event(Event::<T>::CreateAgent { location: Box::new(location), agent_id });
 			Ok(())
 		}
+
+		/// Sends a message to the Gateway contract to create a new Channel representing `origin`
+		///
+		/// - `origin`: Must be `MultiLocation`
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::create_channel())]
+		pub fn create_channel(origin: OriginFor<T>) -> DispatchResult {
+			let location: MultiLocation = T::ControlOrigin::ensure_origin(origin)?;
+
+			let (agent_id, some_para_id, _) = Self::convert_location(location)?;
+
+			ensure!(Agents::<T>::contains_key(agent_id), Error::<T>::AgentNotExist);
+
+			let para_id = some_para_id.ok_or(Error::<T>::LocationToParaIdConversionFailed)?;
+
+			ensure!(!Channels::<T>::contains_key(para_id), Error::<T>::ChannelAlreadyCreated);
+
+			Channels::<T>::insert(para_id, agent_id);
+
+			let message = Message {
+				origin: T::OwnParaId::get(),
+				command: Command::CreateChannel { agent_id, para_id },
+			};
+			Self::submit_outbound(message)?;
+
+			Self::deposit_event(Event::<T>::CreateChannel { para_id, agent_id });
+
+			Ok(())
+		}
+
+		/// Sends a message to the Gateway contract to update channel
+		///
+		/// - `origin`: Must be `MultiLocation`
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::update_channel())]
+		pub fn update_channel(
+			origin: OriginFor<T>,
+			mode: OperatingMode,
+			fee: u128,
+			reward: u128,
+		) -> DispatchResult {
+			let location: MultiLocation = T::ControlOrigin::ensure_origin(origin)?;
```

Comment:
As per my most recent comment in the thread above, we need to ensure that only the origin `/Parachain(X)` can call update channel.

Reply:
Gotcha! It's better to add such limit in `ControlOrigin`.